File: hw/vizcaino/u2/hw14MongoDb/hw14MongoDBPy/view/backup_panel.py
import tkinter as tk
from tkinter import ttk

from dao.backup_dao import BackupDAO
from model.backup import Backup


class BackupPanel(tk.Frame):

    def __init__(self, parent):
        super().__init__(parent)

        self.dao = BackupDAO()

        tk.Label(self, text="ID").grid(row=0, column=0)
        self.txt_id = tk.Entry(self)
        self.txt_id.grid(row=0, column=1)

        tk.Label(self, text="File Name").grid(row=1, column=0)
        self.txt_file_name = tk.Entry(self)
        self.txt_file_name.grid(row=1, column=1)

        tk.Label(self, text="Status").grid(row=2, column=0)
        self.txt_status = tk.Entry(self)
        self.txt_status.grid(row=2, column=1)

        tk.Label(self, text="Date").grid(row=3, column=0)
        self.txt_date = tk.Entry(self)
        self.txt_date.grid(row=3, column=1)

        self.table = ttk.Treeview(
            self,
            columns=("ID", "File Name", "Status", "Date"),
            show="headings"
        )

        for col in ("ID", "File Name", "Status", "Date"):
            self.table.heading(col, text=col)

        self.table.grid(row=4, column=0, columnspan=4, sticky="nsew")

        tk.Button(self, text="Add", command=self.add_backup).grid(row=5, column=0)
        tk.Button(self, text="Delete", command=self.delete_backup).grid(row=5, column=1)
        tk.Button(self, text="Refresh", command=self.load_data).grid(row=5, column=2)

        self.load_data()

    def add_backup(self):

        backup = Backup(
            int(self.txt_id.get()),
            self.txt_file_name.get(),
            self.txt_status.get(),
            self.txt_date.get()
        )

        self.dao.save(backup)

        self.load_data()
        self.clear_fields()

    def delete_backup(self):

        selected = self.table.selection()

        if selected:

            values = self.table.item(selected[0])["values"]

            self.dao.delete(int(values[0]))

            self.load_data()

    def load_data(self):

        for item in self.table.get_children():
            self.table.delete(item)

        for backup in self.dao.find_all():

            self.table.insert(
                "",
                tk.END,
                values=(
                    backup["id"],
                    backup["fileName"],
                    backup["status"],
                    backup["date"]
                )
            )

    def clear_fields(self):

        self.txt_id.delete(0, tk.END)
        self.txt_file_name.delete(0, tk.END)
        self.txt_status.delete(0, tk.END)
        self.txt_date.delete(0, tk.END)

File: hw/vizcaino/u2/hw14MongoDb/hw14MongoDbJava/hw14Interface/src/main/java/ec/edu/espe/view/BackupPanel.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.view;

/**
 *
 * @author Adrian Vizcaino <The-Softwarriors at ESPE>
 */


import ec.edu.espe.model.dao.BackupDAO;
import ec.edu.espe.model.Backup;

import java.awt.BorderLayout;
import java.awt.GridLayout;
import java.util.List;

import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTable;
import javax.swing.JTextField;
import javax.swing.table.DefaultTableModel;

public class BackupPanel extends JPanel {

    private JTextField txtId;
    private JTextField txtFileName;
    private JTextField txtStatus;
    private JTextField txtDate;

    private JTable table;
    private DefaultTableModel model;

    private BackupDAO dao;

    public BackupPanel() {

        dao = new BackupDAO();

        setLayout(new BorderLayout());

        JPanel inputPanel = new JPanel(new GridLayout(4, 2));

        inputPanel.add(new JLabel("ID"));
        txtId = new JTextField();
        inputPanel.add(txtId);

        inputPanel.add(new JLabel("File Name"));
        txtFileName = new JTextField();
        inputPanel.add(txtFileName);

        inputPanel.add(new JLabel("Status"));
        txtStatus = new JTextField();
        inputPanel.add(txtStatus);

        inputPanel.add(new JLabel("Date"));
        txtDate = new JTextField();
        inputPanel.add(txtDate);

        add(inputPanel, BorderLayout.NORTH);

        model = new DefaultTableModel();
        model.addColumn("ID");
        model.addColumn("File Name");
        model.addColumn("Status");
        model.addColumn("Date");

        table = new JTable(model);

        add(new JScrollPane(table), BorderLayout.CENTER);

        JPanel buttonPanel = new JPanel();

        JButton btnAdd = new JButton("Add");
        JButton btnDelete = new JButton("Delete");
        JButton btnRefresh = new JButton("Refresh");

        buttonPanel.add(btnAdd);
        buttonPanel.add(btnDelete);
        buttonPanel.add(btnRefresh);

        add(buttonPanel, BorderLayout.SOUTH);

        btnAdd.addActionListener(e -> addBackup());
        btnDelete.addActionListener(e -> deleteBackup());
        btnRefresh.addActionListener(e -> loadData());

        loadData();
    }

    private void addBackup() {

        try {

            Backup backup = new Backup(
                    Integer.parseInt(txtId.getText()),
                    txtFileName.getText(),
                    txtStatus.getText(),
                    txtDate.getText());

            dao.save(backup);

            loadData();

            clearFields();

        } catch (Exception e) {

            System.out.println("Error: " + e.getMessage());

        }
    }

    private void deleteBackup() {

        int row = table.getSelectedRow();

        if (row >= 0) {

            int id = Integer.parseInt(
                    model.getValueAt(row, 0).toString());

            dao.delete(id);

            loadData();
        }
    }

    private void loadData() {

        model.setRowCount(0);

        List<Backup> backups = dao.findAll();

        for (Backup backup : backups) {

            model.addRow(new Object[]{
                backup.getId(),
                backup.getFileName(),
                backup.getStatus(),
                backup.getDate()
            });
        }
    }

    private void clearFields() {

        txtId.setText("");
        txtFileName.setText("");
        txtStatus.setText("");
        txtDate.setText("");
    }
}